docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,391 @@
+
+When it’s time to remove all related placement rules, not only those rules defined on the table should be removed, but also rules defined on its partitions and indexes.
+
+Once the table is truncated, the placement rules on the table should be updated. Since table id is updated after truncating, key range is also changed.
+
+Since the table can be recovered later by `FLASHBACK` statement, a snapshot of the original placement rules must be saved temporarily. After recovering, the table name is changed, but the table id is the original one, so the snapshot of the original placement rules can be recovered directly.
+
+DDL on partitions and indexes will be discussed below, and other DDL on tables won’t affect placement rules:
+
+* Altering columns
+* Renaming tables
+* Change charset and collation
+
+### DDL on partitions
+
+TiDB supports adding and dropping partitions.
+
+Once a partition is added, its placement rule is empty and the partition follows the rule of the table it belongs to.
+
+Once a partition is dropped, it can’t be recovered anymore, so its placement rules can be removed immediately.
+
+### DDL on indexes
+
+Once a new index is created on an unpartitioned table, the index should follow the rule of the table it belongs to.
+
+Once a new index is created on a table with partitions, each part of the index should follow the rule of the partition it belongs to.
+
+Once an index is dropped, it can’t be recovered anymore, so its placement rules can be removed immediately.
+
+Altering primary index is the same with altering secondary indexes. Because if a primary index can be created or dropped, it must be an unclustered index.
+
+Other DDL on indexes won’t affect placement rules:
+
+* Renaming index
+* Change the visibility of index
+
+### Show DDL jobs
+
+As mentioned earlier, all statements related to placement rules must wait until PD returns. If the execution is interrupted, it should retry until it succeeds or is cancelled, just like other DDL jobs.
+
+PD schedules regions asynchronously after it returns the message. It’s not possible to check the progress of the scheduling for now.
+
+Ongoing and finished statements related to placement rules can also be queried through `ADMIN SHOW DDL`, `ADMIN SHOW DDL JOBS`, or other similar statements.
+
+## View rules
+
+All placement rules should be queried through statements.
+
+### System table
+
+A new system table `information_schema.placement_rules` can be added to view all placement rules. The table contains such columns:
+
+* rule_id
+* database_id and database_name
+* table_id and table_name
+* partition_id and partition_name
+* index_id and index_name
+* label
+* role
+* count
+
+The system table is a virtual table, which doesn’t store data. When querying the table, TiDB queries PD and integrates the result in a table format. That also means the metadata is stored on PD instead of TiKV.
```

Comment:
Is it possible to return the placement rules defined through `pd-ctl`?

Reply:
Yes, I think. The current implementation is using PD api the same way as `pd-ctl`.